proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+The "SwiftAsync" calling convention has an additional "async context" register and specific behaviors expected around how the Swift async state machine is called. Full investigation into the differences needed to implement it are still TBD.
+
+##### Self register
+
+We have two options for supporting the Self register in the calling convention:
+
+1. Use an attribute like `[SwiftSelf]` on a parameter to specify it should go in the self register.
+2. Specify that combining the `Swift` calling convention with the `MemberFunction` calling convention means that the first argument is the `self` argument.
+
+The first option seems like a natural fit, but there is one significant limitation: Attributes cannot be used in function pointers. Function pointers are a vital scenario for us to support as we want to support virtual method tables as they are used in scenarios like protocol witness tables. The second option is a natural fit as the `MemberFunction` calling convention combined with the various C-based calling conventions specifies that there is a `this` argument as the first argument. Defining `Swift` + `MemberFunction` to imply/require the `self` argument is a perfect conceptual extension of the existing model.
+
+###### Error register
+
+We have many options for handling the error register in the Swift calling convention:
+
+1. Use an attribute like `[SwiftError]` on a by-ref or `out` parameter to indicate the error parameter.
+2. Use a special type like `SwiftError<T>` on a by-ref or `out` parameter to indicate the error parameter.
+3. Use a special return type `SwiftReturn<T, TError>` to indicate the error parameter and combine it with the return value.
+4. Use special helper functions `Marshal.Get/SetLastSwiftError` to get and set the last Swift error. Our various compilers will automatically emit a call to `SetLastSwiftError` from Swift functions and emit a call to `GetLastSwiftError` in the epilog of `UnmanagedCallersOnly` methods. The projection generation will emit calls to `Marshal.Get/SetLastSwiftError` to get and set these values to and from exceptions.
+
+We have a prototype that uses the first option; however, using an attribute has the same limitations as the attribute option for the self register. As a result, we should consider an alternative option.
+
+Options 2 and 3 have similar characteristics: They both express that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. Option 2 is likely cheaper, as the argument with `ref/out SwiftError<T>` type can easily be associated with the error register. With Option 2 though, we'd have to determine the behavior if multiple parameters used the `SwiftError` type (likely throw an exception). Option 3 provides a more functional-programming-style API that is more intuitive, but likely more expensive to implement.
```

Comment:
> Option 3 provides a more functional-programming-style API that is more intuitive, but likely more expensive to implement.

@stephen-hawley Is there a practical use case for the `SwiftReturn<T, TError>`? This may be responsibility of the projection tools.

Reply:
I wouldn't a method like that. I would write a C# method that takes does something like this:
```
public nint SomeMethodThatMightThrow ()
{
    var errReturn = SomeMethodThatMightThrow_Pinvoke ();
    if (errReturn.HasError) {
        throw SwiftException.FromError (errReturn.Error);
    }
    return errReturn.Value;
}
```

Reply:
But also important to consider is the other side of the coin: C# overrides an `open` class from swift and overrides and `open` method that throws. Today, BTfS does this for all `open` classes:
1. write a swift class that overrides ever overridable method which includes a simulated vtable for virtual methods
2. write a C# class named after the original class which in its static constructor sends a pile of function pointers into the swift class which are used to populate the vtable.  Each of the function pointers points to a static C# method which I call a receiver. The receiver gets (among other things) the swift instance which it uses to look up the C# instance to handle the call and a pointer to a Medusa tuple for the return and exception.
3. In the case of an exception, BTfS injects a try catch, projects the exception onto a swift exception. This in turns gets set into the Medusa tuple and sent back to swift where the invokes unmarshals it.
This is a lot of work and I would like to find ways for it to be more betterer.

Reply:
I should define this - a Medusa tuple is of the form `(Bool, T, Error)` where T is the return type of a function. A method that may throw gets passed a pointer to memory allocated for a Medusa tuple, but leaves it uninitialized. If it has thrown, it sets the `Bool` portion to true and sets the `Error` portion leaving `T` uninitialized. In the case of not throwing, it sets the `Bool` portion to false and sets the `T` portion. This violates swift semantics in that Swift *really* hates having uninitialized members, and if you ever try to look at the tuple, you will die. Why not use a Swift enum for this and crack it open? Because getting at the payload of an enum in C# is roughly 5-10x the cost of getting at the elements of an tuple and the means of accessing a tuple's elements can be memoized whereas getting the payload out of an enum is fixed and can't be memoized in a useful way.
When I started looking at exceptions in Swift, I was surprised that they didn't just define them like this:
```
public enum ErrorReturn<T> {
    case value (T)
    case error(Error)
}
```
and make the call syntactic sugar. If I had to speculate, it was to make interoperation with ObjC errors easier/fast/smoother/more bettererer.

Reply:
One thought -- this is very much expressing exceptions in Swift. The syntax is even throw/catch. Should we simply expose this as a C# exception from the runtime? Maybe of a specific type, like a new SwiftException class?

Reply:
Exposing the error directly as an exception from the runtime layer is what we did in Objective-C. It introduces additional complexity in the interop layer (we always generate IL stubs for P/Invokes into the Objective-C selector functions to enable this feature) and it doesn't enable a projection to provide a nice error-mapping experience (ie to translate a swift error for "file not found" into FileNotFoundException would require a try-catch-throw instead of just an "if error then throw" flow).

Reply:
Two follow-up questions: the Swift ABI describes that throwing is a part of the signature, but there's effectively a covariance relationship in that non-throwing functions can convert to throwing versions. If we go the extra-parameter route, presumably it would be up to either the user or the translation layer to provide a conversion function which never sets the error.

And second, what happens for exception thrown in C# code that is called from Swift? Kind of the same question of what happens in C++, but I could see us wanting to provide a slightly better experience for Swift than for C++.

Reply:
> I could see us wanting to provide a slightly better experience for Swift than for C++.

We have some mechanism for Objective-C on this front, but it is global and less than ideal. It also requires the ownership platform (in this case, Objective-C) to provide a mechanism for throwing an exception in their desired manner. I would be very reluctant to try and handle the throwing of a Swift exception from the runtime ourselves or define it in managed code.

All that aside, if all that is needed to "throw" is creating a Swift projection and setting the error pointer, then that should be relatively simple.